// backend/src/modules/user/user.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../../prisma/prisma.service';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { SyncUserDto } from './dto/sync-user.dto';
import type { SyncUserResponse } from '@ai-media-bot/contracts';

@Injectable()
export class UserService {
  constructor(private readonly prisma: PrismaService) {}

  async create(createUserDto: CreateUserDto) {
    return await this.prisma.user.create({
      data: {
        telegramId: BigInt(createUserDto.telegramId),
        username: createUserDto.username,
        firstName: createUserDto.firstName,
      },
    });
  }

  async update(id: string, updateUserDto: UpdateUserDto) {
    const data: any = {
      username: updateUserDto.username,
      firstName: updateUserDto.firstName,
    };

    if (updateUserDto.telegramId) {
      data.telegramId = BigInt(updateUserDto.telegramId);
    }

    return await this.prisma.user.update({
      where: { id },
      data: data,
    });
  }

  async findAll() { return await this.prisma.user.findMany(); }
  async findOne(id: string) { return await this.prisma.user.findUnique({ where: { id } }); }
  async remove(id: string) { return await this.prisma.user.delete({ where: { id } }); }

  // --- Task 1.3: User Sync (Telegram-specific, sesuai deviation log) ---
  async syncUser(dto: SyncUserDto): Promise<SyncUserResponse> {
    const telegramId = BigInt(dto.telegramId);

    const user = await this.prisma.user.upsert({
      where: { telegramId },
      create: {
        telegramId,
        firstName: dto.firstName,
        lastName: dto.lastName,
        username: dto.username,
        languageCode: dto.languageCode,
      },
      update: {
        // Prisma mengabaikan field bernilai `undefined` di objek data —
        // field yang tidak dikirim tidak akan menimpa data lama.
        firstName: dto.firstName,
        lastName: dto.lastName,
        username: dto.username,
        languageCode: dto.languageCode,
      },
    });

    return this.toSyncUserResponse(user);
  }

  private toSyncUserResponse(user: {
    id: string;
    telegramId: bigint;
    firstName: string | null;
    lastName: string | null;
    username: string | null;
    languageCode: string | null;
    creditBalance: number;
    createdAt: Date;
    updatedAt: Date;
  }): SyncUserResponse {
    return {
      id: user.id,
      telegramId: user.telegramId.toString(),
      firstName: user.firstName ?? undefined,
      lastName: user.lastName ?? undefined,
      username: user.username ?? undefined,
      languageCode: user.languageCode ?? undefined,
      creditBalance: user.creditBalance,
      createdAt: user.createdAt.toISOString(),
      updatedAt: user.updatedAt.toISOString(),
    };
  }
}
